File: src/189-rotate-array.rs
// 189. Rotate Array
// https://leetcode.com/problems/rotate-array/

struct Solution;

impl Solution {
    pub fn rotate(nums: &mut Vec<i32>, k: i32) {
        let k = nums.len() - k as usize % nums.len();
        *nums = nums[k..].iter().chain(nums[..k].iter()).copied().collect()
    }
}

#[rustfmt::skip]
fn main() {
    [
        (vec![1, 2, 3, 4, 5, 6, 7], 3),
        (vec![1, 2, 3, 4, 5, 6, 7], 100_000),
        (vec![-1, -100, 3, 99], 2),
        (vec![-1], 20),
    ]
    .iter_mut()
    .for_each(|(nums, k)| {
        Solution::rotate(nums, *k);
        println!("{:?}", nums);
    })
}

File: src/c/189-rotate-array.c
// 189. Rotate Array
// https://leetcode.com/problems/rotate-array/

#include <stdio.h>

void rotateSingle(int* nums, int numsSize) {
    int temp = nums[numsSize - 1];
    for (int i = 0; i < numsSize - 1; ++i)
        nums[numsSize - 1 - i] = nums[numsSize - 2 - i];
    nums[0] = temp;
}

void rotate(int* nums, int numsSize, int k) {
    k = k % numsSize;
    for (int i = 0; i < k; ++i)
        rotateSingle(nums, numsSize);
}

int main() {
    int nums[] = {1,2,3,4,5,6,7};
    rotate(nums, 7, 3);
    for (int i = 0; i < 7; ++i) printf("%d,", nums[i]);
    printf("\n");
}
